### PR TITLE
crd/ContourDeployment: Add fields 'extraVolumes & extraVolumeMounts'

### PR DESCRIPTION
**Pull request by izturn**
add the fields `extraVolumes` &`extraVolumeMounts` to `crd/ContourDeployment` to let `pod/envoy` can mount  more volumes

Signed-off-by: Gang Liu [gang.liu@daocloud.io](mailto:gang.liu@daocloud.io)

### CODE REVIEW

**Comment by izturn** [comment]
ping @skriss  @tsaarni

**Comment by izturn** [comment]
ping again @skriss @tsaarni

**Comment by sunjayBhatia** [comment]
@izturn could you add some commentary on what kinds of additional things you expect users will want to mount into the Envoy pod? I don't know if we've had requests to do so before, but our basic example is just YAML so people can do as they please

**Comment by izturn** [comment]
with my use case, I want to pull the `envoy_access.log` at the specific path, such as `/tmp` by `elk`, so adding this feature will help that

File: internal/provisioner/objects/dataplane/dataplane_test.go [hunk 1]
```diff
@@ -125,6 +125,36 @@ func checkDaemonSetHasNodeSelector(t *testing.T, ds *appsv1.DaemonSet, expected
 	t.Errorf("deployment has unexpected node selector %q", expected)
 }
 
+func checkDaemonSetHasVolume(t *testing.T, ds *appsv1.DaemonSet, vol corev1.Volume, volMount corev1.VolumeMount) {
```
Comment (sunjayBhatia) [inline internal/provisioner/objects/dataplane/dataplane_test.go:128]: if we end up keeping the init container modification, should test that here

File: internal/provisioner/objects/dataplane/dataplane.go [hunk 2]
```diff
@@ -307,6 +307,13 @@ func desiredContainers(contour *model.Contour, contourImage, envoyImage string)
 		},
 	}
 
+	for i := range initContainers {
```
[the remark this answers is not in the document]
Comment (izturn) [reply internal/provisioner/objects/dataplane/dataplane.go:310]: it seems  we should not add these to init containers, WDYT @skriss ?

File: changelogs/unreleased/4680-izturn-small.md [hunk 3]
```diff
@@ -0,0 +1 @@
+add the fields extraVolumes & extraVolumeMounts to crd/ContourConfiguration to let pod/envoy can mount more volumes
```
Comment (skriss) [inline changelogs/unreleased/4680-izturn-small.md:1]:
```suggestion
add the fields extraVolumes & extraVolumeMounts to crd/ContourDeployment to enable Envoy pods to mount additional volumes
```

File: internal/provisioner/objects/dataplane/dataplane.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (skriss) [reply internal/provisioner/objects/dataplane/dataplane.go:310]: Agreed, we don't need to add them to the init containers.

File: apis/projectcontour/v1alpha1/contourdeployment.go [hunk 4]
```diff
@@ -123,6 +123,14 @@ type EnvoySettings struct {
...
 	// +optional
 	NodePlacement *NodePlacement `json:"nodePlacement,omitempty"`
+
+	// ExtraVolumes holds the extra volumes to add.
+	// +optional
+	ExtraVolumes []corev1.Volume `json:"extraVolumes,omitempty"`
+
+	// ExtraVolumeMounts holds the extra volume mounts to add (normally used with extraVolumes).
+	// +optional
+	ExtraVolumeMounts []corev1.VolumeMount `json:"extraVolumeMounts,omitempty"`
```
Comment (skriss) [inline apis/projectcontour/v1alpha1/contourdeployment.go:133]: I think we should add validation logic to the GatewayClass controller within the provisioner, to ensure that any ExtraVolumeMounts reference a volume from ExtraVolumes -- does that make sense? (In here: https://github.com/projectcontour/contour/blob/main/internal/provisioner/controller/gatewayclass.go#L159)

Comment (sunjayBhatia) [reply apis/projectcontour/v1alpha1/contourdeployment.go:133]: yep good idea +1

Comment (izturn) [reply apis/projectcontour/v1alpha1/contourdeployment.go:133]: i will upgrade it later

File: internal/provisioner/controller/gatewayclass.go [hunk 5]
```diff
@@ -181,6 +181,19 @@ func (r *gatewayClassReconciler) Reconcile(ctx context.Context, req ctrl.Request
 					invalidParamsMessages = append(invalidParamsMessages, msg)
 				}
 			}
+
+			if params.Spec.Envoy.ExtraVolumeMounts != nil {
```
Comment (skriss) [inline internal/provisioner/controller/gatewayclass.go:185]: Please add a test case that covers this new code branch, similar to https://github.com/projectcontour/contour/blob/main/internal/provisioner/controller/gatewayclass_test.go#L243-L277

**Comment by skriss** [comment]
@izturn once you've added the requested test case, I think this is ready for merge.

File: internal/provisioner/controller/gatewayclass_test.go [hunk 6]
```diff
@@ -266,6 +268,16 @@ func TestGatewayClassReconcile(t *testing.T) {
...
+						ExtraVolumeMounts: []corev1.VolumeMount{
+							{
+								Name: "volume-a",
+							},
+						},
+						ExtraVolumes: []corev1.Volume{
+							{
+								Name: "volume-b",
+							},
+						},
```
Comment (skriss) [inline internal/provisioner/controller/gatewayclass_test.go:280]: This should go in a separate test case - the case you edited is testing for an invalid NetworkPublishing type. Please create a separate one specifically for ExtraVolumes/ExtraVolumeMounts.

**Review by sunjayBhatia — APPROVED** [review]

**Pull request merged**